File: internal/risk/service.go
package risk

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"log/slog"
	"time"

	"github.com/jackc/pgx/v5"
	"github.com/nats-io/nats.go/jetstream"

	"github.com/noeljackson/stocks/internal/platform/bus"
	"github.com/noeljackson/stocks/internal/platform/store"
	"github.com/noeljackson/stocks/internal/platform/subjects"
)

// actionable mirrors the JSON shape published by the (forthcoming) thesis
// engine on thesis.actionable. v0 uses the minimal subset the risk overlay
// needs — the engine is free to add fields without breaking us.
type actionable struct {
	ThesisID      string  `json:"thesis_id"`
	Symbol        string  `json:"symbol"`
	Cluster       string  `json:"cluster"`
	Instrument    string  `json:"instrument"`
	DeltaNotional float64 `json:"delta_notional"`
	PremiumAtRisk float64 `json:"premium_at_risk"`
}

// portfolioDefaults provides the cash%/drawdown% that v0 can't yet compute
// from a real broker integration. Override via env if you're paper-trading
// against a known balance (TODO: wire into Service.PortfolioOverride).
var portfolioDefaults = Portfolio{TotalValue: 100_000, CashPct: 50, DrawdownPct: 0}

type Service struct {
	DB                *store.DB
	Bus               *bus.Bus
	Log               *slog.Logger
	PortfolioOverride *Portfolio // optional; nil → use computed/default
}

func New(db *store.DB, b *bus.Bus, log *slog.Logger) *Service {
	return &Service{DB: db, Bus: b, Log: log}
}

func (s *Service) Run(ctx context.Context) error {
	if err := s.Bus.EnsureStream(ctx, subjects.StreamDecisions, "risk.*", "decision.*"); err != nil {
		return fmt.Errorf("ensure DECISIONS: %w", err)
	}
	stop, err := s.Bus.Consume(ctx, subjects.StreamThesis, "risk-overlay", "thesis.actionable",
		func(m jetstream.Msg) error { return s.onActionable(ctx, m) })
	if err != nil {
		return err
	}
	defer stop()
	s.Log.Info("risk overlay consuming", "stream", subjects.StreamThesis, "filter", "thesis.actionable")
	<-ctx.Done()
	return nil
}

func (s *Service) onActionable(ctx context.Context, m jetstream.Msg) error {
	var a actionable
	if err := json.Unmarshal(m.Data(), &a); err != nil {
		s.Log.Warn("dropping malformed thesis.actionable", "err", err)
		return nil // ack-and-drop
	}
	cfg, ver, err := s.loadConfig(ctx)
	if err != nil {
		s.Log.Error("load risk config", "err", err)
		return err
	}
	positions, err := s.loadOpenPositions(ctx)
	if err != nil {
		s.Log.Error("load positions", "err", err)
		return err
	}
	port := portfolioDefaults
	if s.PortfolioOverride != nil {
		port = *s.PortfolioOverride
	}

	decision := Evaluate(toIntent(a), positions, port, cfg)
	if !decision.Veto && len(decision.Warnings) == 0 {
		return nil // clean: nothing to publish
	}

	payload, _ := json.Marshal(map[string]any{
		"thesis_id":      a.ThesisID,
		"symbol":         a.Symbol,
		"veto":           decision.Veto,
		"reasons":        decision.Reasons,
		"warnings":       decision.Warnings,
		"size_mult":      decision.SizeMult,
		"config_version": ver,
		"at":             time.Now().UTC(),
	})
	subj := subjects.RiskWarning
	if decision.Veto {
		subj = subjects.RiskVeto
	}
	if err := s.Bus.Publish(ctx, subj, payload); err != nil {
		s.Log.Error("publish risk verdict", "subject", subj, "err", err)
		return err
	}
	s.Log.Info("risk verdict", "subject", subj, "symbol", a.Symbol,
		"veto", decision.Veto, "reasons", decision.Reasons, "warnings", decision.Warnings)
	return nil
}

func (s *Service) loadConfig(ctx context.Context) (Config, int, error) {
	body, ver, err := s.DB.ActiveConfig(ctx, "risk")
	if err != nil {
		return Config{}, 0, err
	}
	var c Config
	if err := json.Unmarshal(body, &c); err != nil {
		return Config{}, 0, err
	}
	return c, ver, nil
}

func (s *Service) loadOpenPositions(ctx context.Context) ([]Position, error) {
	rows, err := s.DB.Pool.Query(ctx, `
		SELECT p.symbol, COALESCE(t.cluster_id, ''), p.instrument,
		       COALESCE(p.delta_notional, 0), COALESCE(p.premium_at_risk, 0)
		  FROM position p
		  LEFT JOIN ticker t ON t.symbol = p.symbol
		 WHERE p.closed_at IS NULL`)
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			return nil, nil
		}
		return nil, err
	}
	defer rows.Close()
	var out []Position
	for rows.Next() {
		var p Position
		if err := rows.Scan(&p.Symbol, &p.Cluster, &p.Instrument, &p.DeltaNotional, &p.PremiumAtRisk); err != nil {
			return nil, err
		}
		out = append(out, p)
	}
	return out, rows.Err()
}

func toIntent(a actionable) Intent {
	return Intent{
		Symbol:        a.Symbol,
		Cluster:       a.Cluster,
		Instrument:    a.Instrument,
		DeltaNotional: a.DeltaNotional,
		PremiumAtRisk: a.PremiumAtRisk,
	}
}

File: src/risk/service.rs
//! Risk overlay service: durable consumer on thesis.actionable, loads
//! positions from Postgres, publishes risk.veto / risk.warning.

use std::sync::Arc;

use anyhow::Result;
use chrono::Utc;
use serde::Deserialize;
use tracing::{info, warn};

use super::{Config, Intent, Portfolio, derive_portfolio, evaluate};
use crate::platform::bus::{Bus, ConsumerHandle};
use crate::platform::store::Store;
use crate::platform::subjects;

#[derive(Deserialize, Default)]
struct Actionable {
    #[serde(default)]
    thesis_id: String,
    #[serde(default)]
    symbol: String,
    #[serde(default)]
    cluster: String,
    #[serde(default)]
    instrument: String,
    #[serde(default)]
    delta_notional: f64,
    #[serde(default)]
    premium_at_risk: f64,
}

// Honest fallback when the operator hasn't set `portfolio_settings` yet.
// Marked explicitly so risk verdicts emitted during this state can be
// recognized as demo-mode (we also log a warn! at every evaluation).
const UNCONFIGURED_PORTFOLIO: Portfolio = Portfolio {
    total_value: 100_000.0,
    cash_pct: 50.0,
    drawdown_pct: 0.0,
};

pub async fn run(store: Store, bus: Bus) -> Result<ConsumerHandle> {
    bus.ensure_stream(subjects::STREAM_DECISIONS, &["risk.*", "decision.*"])
        .await?;

    let store = Arc::new(store);
    let bus = Arc::new(bus);
    let bus_consume = bus.clone();
    let handle = bus_consume
        .consume(
            subjects::STREAM_THESIS,
            "risk-overlay",
            subjects::THESIS_ACTIONABLE,
            {
                let store = store.clone();
                let bus = bus.clone();
                move |msg| {
                    let store = store.clone();
                    let bus = bus.clone();
                    async move { on_actionable(&store, &bus, &msg.payload).await }
                }
            },
        )
        .await?;
    info!(
        stream = subjects::STREAM_THESIS,
        filter = subjects::THESIS_ACTIONABLE,
        "risk overlay consuming"
    );
    Ok(handle)
}

async fn on_actionable(store: &Store, bus: &Bus, data: &[u8]) -> Result<()> {
    let Ok(a) = serde_json::from_slice::<Actionable>(data) else {
        warn!("dropping malformed thesis.actionable");
        return Ok(());
    };
    let (cfg_json, cfg_ver) = store.active_config("risk").await?;
    let cfg: Config = serde_json::from_value(cfg_json)?;
    let positions = store.open_positions_for_risk().await?;
    let settings = store.portfolio_settings().await?;
    let realized_pnl = store.realized_pnl_total().await.unwrap_or(0.0);

    let (portfolio, is_demo) = match derive_portfolio(settings, &positions, realized_pnl) {
        Some(p) => (p, false),
        None => {
            warn!(
                "portfolio_settings.account_size_usd unset — risk overlay running on \
                 DEMO portfolio (100k notional, 50% cash). Set via PUT /api/portfolio."
            );
            (UNCONFIGURED_PORTFOLIO, true)
        }
    };

    let intent = Intent {
        symbol: a.symbol.clone(),
        cluster: a.cluster.clone(),
        instrument: a.instrument.clone(),
        delta_notional: a.delta_notional,
        premium_at_risk: a.premium_at_risk,
    };
    let decision = evaluate(&intent, &positions, portfolio, &cfg);

    if !decision.veto && decision.warnings.is_empty() {
        return Ok(());
    }

    let payload = serde_json::json!({
        "thesis_id": a.thesis_id,
        "symbol":    a.symbol,
        "veto":      decision.veto,
        "reasons":   decision.reasons,
        "warnings":  decision.warnings,
        "size_mult": decision.size_mult,
        "config_version": cfg_ver,
        "portfolio_demo": is_demo,
        "portfolio": {
            "total_value": portfolio.total_value,
            "cash_pct":    portfolio.cash_pct,
            "drawdown_pct": portfolio.drawdown_pct,
        },
        "at":        Utc::now(),
    });
    let subject = if decision.veto { subjects::RISK_VETO } else { subjects::RISK_WARNING };
    bus.publish(subject, payload.to_string().as_bytes()).await?;
    info!(
        subject = subject,
        symbol = %a.symbol,
        veto = decision.veto,
        "risk verdict"
    );
    Ok(())
}

